Zend/zend_attributes.c

```diff
@@ -69,33 +70,30 @@ uint32_t zend_attribute_attribute_get_flags(zend_attribute *attr, zend_class_ent
 	return ZEND_ATTRIBUTE_TARGET_ALL;
 }
 
-static void validate_allow_dynamic_properties(
+static zend_string *validate_allow_dynamic_properties(
 		zend_attribute *attr, uint32_t target, zend_class_entry *scope)
 {
+	ZEND_ASSERT(scope != NULL);
+	const char *msg = NULL;
 	if (scope->ce_flags & ZEND_ACC_TRAIT) {
-		zend_error_noreturn(E_ERROR, "Cannot apply #[\\AllowDynamicProperties] to trait %s",
-			ZSTR_VAL(scope->name)
-		);
-	}
-	if (scope->ce_flags & ZEND_ACC_INTERFACE) {
-		zend_error_noreturn(E_ERROR, "Cannot apply #[\\AllowDynamicProperties] to interface %s",
-			ZSTR_VAL(scope->name)
-		);
-	}
-	if (scope->ce_flags & ZEND_ACC_READONLY_CLASS) {
-		zend_error_noreturn(E_ERROR, "Cannot apply #[\\AllowDynamicProperties] to readonly class %s",
-			ZSTR_VAL(scope->name)
-		);
+		msg = "Cannot apply #[\\AllowDynamicProperties] to trait %s";
+	} else if (scope->ce_flags & ZEND_ACC_INTERFACE) {
+		msg = "Cannot apply #[\\AllowDynamicProperties] to interface %s";
+	} else if (scope->ce_flags & ZEND_ACC_READONLY_CLASS) {
+		msg = "Cannot apply #[\\AllowDynamicProperties] to readonly class %s";
+	} else if (scope->ce_flags & ZEND_ACC_ENUM) {
+		msg = "Cannot apply #[\\AllowDynamicProperties] to enum %s";
 	}
-	if (scope->ce_flags & ZEND_ACC_ENUM) {
-		zend_error_noreturn(E_ERROR, "Cannot apply #[\\AllowDynamicProperties] to enum %s",
-			ZSTR_VAL(scope->name)
-		);
+	if (msg != NULL) {
+		smart_str str = {0};
+		smart_str_append_printf(&str, msg, ZSTR_VAL(scope->name));
+		return smart_str_extract(&str);
```

Comment:
Could be: `zend_strpprintf()`:
```suggestion
		return zend_strpprintf(0, msg, ZSTR_VAL(scope->name));
```

Reply:
learn something new every day - done